doc/how_to/migrate/from_streamlit.md

```diff
@@ -0,0 +1,950 @@
+
+```python
+@pn.cache
+def plot(data, bins):
+    ...
+```
+
+Be aware the the annotation approach has a tendency to mix up your business logic
+(`data` and `plot` function) and your caching logic (when and how to apply caching). This can make
+your code harder to maintain and reuse for other use cases.
+
+### Global Cache Example
+
+This session will show you how to cache globally, i.e. across user sessions
+
+#### Streamlit Global Cache Example
+
+```python
+from time import sleep
+
+import numpy as np
+import streamlit as st
+from matplotlib.figure import Figure
+
+@st.cache_resource
+def get_data():
+    print("get_data func")
+    sleep(1.0)
+    return np.random.normal(1, 1, size=100)
+
+@st.cache_data(hash_funcs={Figure: lambda _: None})
+def plot(data, bins):
+    print("plot func", bins)
+    sleep(0.25)
+    fig = Figure(figsize=(8,4))
+    ax = fig.subplots()
+    ax.hist(data, bins=bins)
+    return fig
+
+data = get_data()
+bins = st.slider(value=20, min_value=10, max_value=30, step=1, label="Bins")
+st.pyplot(plot(data, bins))
+```
+
+The global cache `st.cache_resource` is used on the `get_data` function to only load the data once
+across all users sessions.
+
+#### Panel Global Cache Example
+
+```python
+from time import sleep
+
+import numpy as np
+import panel as pn
+from matplotlib.figure import Figure
+
+
+def get_data():
+    print("data func")
+    sleep(1.0)
+    return np.random.normal(1, 1, size=100)
+
+def plot(data, bins):
+    print("plot func", bins)
+    sleep(0.25)
+    fig = Figure(figsize=(8,4))
+    ax = fig.subplots()
+    ax.hist(data, bins=bins)
+    return fig
+
+pn.extension(sizing_mode="stretch_width", template="bootstrap")
+
+data = pn.state.as_cached(key="data", fn=get_data) # Global Cache
+bins = pn.widgets.IntSlider(value=20, start=10, end=30, step=1)
+cplot = pn.cache(plot) # Session Cache
+bplot = pn.bind(cplot, data, bins)
+pn.Column(bins, bplot).servable()
+```
+
+The global cache `pn.state.as_cached` is used on the `get_data` function to only load the data once
+across all users sessions.
+
+### Cache Migration Steps
+
+To migrate
+
+- replace `st.cache_data` with `pn.cache` to migrate your *session caching*
+  - You only need to cache expensive, *bound functions*.
+- replace `st.cache_resource` with `pn.state.as_cached` to migrate your *global caching*
+  - You only need to cache expensive, *bound functions* and objects you want to share across user
+  sessions
+
+In the process consider separating your business logic and caching logic
+(when and how to apply caching) to get more maintainable and reusable code.
+
+## Multi Page Apps
+
+Migrating your Streamlit multi page app to Panel is simple. In Panel each page is simply a file
+that you *serve*
+
+```bash
+panel serve home.py page1.py page2.ipynb
+```
+
+You can specify the *home* page with the `--index` flag.
+
+```bash
+panel serve home.py page1.py page2.ipynb --index=home
+```
+
```

Comment:
Something I think you're missing (and missing in my medium article) is how hard it is to dynamically creating layouts and keeping their states.

As an example, in [MusicAIl](https://github.com/ahuang11/musicail/blob/main/app.py#L312-L322) I had quite a headache dealing with session state.

My goal was to create a new pn.Card for each instrument and the notes that each instrument plays, but in Streamlit, you need to manually keep track of the IDs of each widget. In Panel state handling is *so* straightforward so I think it should be highlighted somewhere.

Reply:
I would really like to add a section on "session state" but it was added to Streamlit after I left that framework behind. And it's where I think their promise of a fast way to build data apps stops. They also have callbacks now. But until v0.81 I believe their mantra was "no callbacks required"

Reply:
Could you share some Streamlit example that shows the issue with session state.

Reply:
Yeah I'll try to work on one regarding dynamic layouts.

Reply:
I resolve this one and add a todo item to the todo list in the first post.